multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java

```diff
@@ -224,8 +221,24 @@ private CloudServiceInstanceExtended mergeCredentials(CloudServiceInstanceExtend
                                                     .withCredentials(result);
     }
 
-    private boolean shouldUpdateKeys(CloudServiceInstanceExtended service, List<CloudServiceKey> serviceKeys) {
-        return !(service.isUserProvided() || CollectionUtils.isEmpty(serviceKeys));
+    private boolean shouldUpdateKeys(CloudServiceInstanceExtended service, CloudServiceInstance existingService, ProcessContext context) {
+        if (service.isUserProvided()) {
+            return false;
+        }
+        return serviceHasServiceKeysToCreate(context, service) || shouldDeleteServiceKeys(existingService, context);
+    }
+
+    private boolean serviceHasServiceKeysToCreate(ProcessContext context, CloudServiceInstanceExtended service) {
```

Comment:
I don't think we need "service" as a prefix here.
You can omit it and name the method `hasServiceKeysForCreation`.